src/herder/TxSetFrame.cpp

```diff
@@ -33,43 +34,51 @@ namespace stellar
 
 using namespace std;
 
+// Construct a valid TxSetFrame from a list of transactions (e.g. from the
+// TransactionQueue). We make sure the transactions are sorted in (and
+// maintains) hash order and computes the contents hash on contruction.
+TxSetFrame::TxSetFrame(Hash const& previousLedgerHash,
+                       Transactions const& transactions)
+    : mPreviousLedgerHash(previousLedgerHash)
+    , mTxs(TxSetUtils::sortTxsInHashOrder(transactions))
+    , mTxsInValidHashOrder(true)
+    , mHash(TxSetUtils::computeContentsHash(previousLedgerHash, mTxs))
+{
+}
+
 TxSetFrame::TxSetFrame(Hash const& previousLedgerHash)
-    : mHash(std::nullopt)
-    , mValid(std::nullopt)
-    , mPreviousLedgerHash(previousLedgerHash)
+    : TxSetFrame(previousLedgerHash, Transactions{})
 {
 }
 
+// Construct a TxSetFrame from an xdrSet, which may directly come from the
+// overlay and be invalid (in the incorrect order), in which case we mark
+// this set invalid and do not compute its hash.
 TxSetFrame::TxSetFrame(Hash const& networkID, TransactionSet const& xdrSet)
-    : mHash(std::nullopt), mValid(std::nullopt)
+    : mPreviousLedgerHash(xdrSet.previousLedgerHash)
+    , mTxs(TxSetUtils::extractTxsFromXdrSet(networkID, xdrSet))
+    , mTxsInValidHashOrder(TxSetUtils::isInValidHashOrder(mTxs))
```

Comment:
nit: just realized that the name `mTxsInValidHashOrder` kinda reads too close to `mTxsInvalidHashOrder`, so we should make it `mTxsIsValidHashOrder` (same with `TxSetUtils::isInValidHashOrder` that should be `TxSetUtils::isValidHashOrder`)